File: UsingPython/subsets.cc
#include <iostream>
#include <vector>

using namespace std;

void subsets(vector<int>& array, size_t index, vector<vector<int> >& res, vector<int>& pRes)
{
  if (index == array.size()) {
    res.push_back(pRes);
    return;
  }

  subsets(array, index+1, res, pRes);
  pRes.push_back(array[index]);
  subsets(array, index+1, res, pRes);
  pRes.pop_back();
}

int main()
{
  vector<int> array = {1,2,3};
  vector<vector<int> > res;
  vector<int> pRes;
  subsets(array, 0, res, pRes);

  for (auto a: res) {
    cout << "{ ";
    for (auto e: a) {
      cout << e << ",";
    }
    cout << "}";
  }
  cout << endl;

  return 0;
}
  

File: UsingPython/subsets.py
"""
All subsets of a set
"""

def subsets(arr, index, res, pRes):
    print('index:{}, res: {}, pRes; {}\n'.format(index, res, pRes))
    if index == len(arr):
        res.append(pRes)
        return

    subsets(arr, index+1, res, pRes[:])
    pRes.extend([arr[index]])
    subsets(arr, index+1, res, pRes[:])
    pRes.pop()


def main():
    res = []
    pRes = []
    subsets([1,2], 0, res, pRes)
    print res


if __name__ == "__main__":
    main()
